### PR TITLE
TODO 15

### PR DESCRIPTION
**Pull request by arkadiuszpietraszek**
write about:
setReadValueRequestCallback() for characteristic
setWriteValueRequestCallback() for characteristic
setReadValueRequestCallback() for descriptor
setWriteValueRequestCallback() for descriptor

### CODE REVIEW

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 1]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+      return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;
+   };
+
+   server.services[0].characteristics[0].setWriteValueRequestCallback(characteristicWriteRequestCallback);
+   ```
+
+   > **Note**
+   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this characteristic.
+
+6. To register the callback called when a client reads the value of the desriptor from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the desriptor object.
```
Comment (djuszczakk) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
6. To register the callback called when a client reads the value of the descriptor from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the descriptor object.
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 2]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+      console.log(clientAddress + " requested to read desriptor's value with offset: " + offset);
+      return new tizen.GATTRequestReply(0, "0x1234");
+   };
+   server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(characteristicReadRequestCallback);
+   ```
+
+   > **Note**
+   > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` ot this descriptor.
+
+7. To register the callback called when a client writes a value of the desriptor of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the desriptor object.
```
Comment (djuszczakk) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
7. To register the callback called when a client writes a value of the descriptor of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the descriptor object.
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 3]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+      console.log(clientAddress + " requested to write desriptor's value: " + value +
+                  " with offset: " + offset);
+      return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;
+   };
+
+   server.services[0].characteristics[0].descriptors[0].setWriteValueRequestCallback(desriptorWriteRequestCallback);
+   ```
+
+   > **Note**
+   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this desriptor.
```
Comment (djuszczakk) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this descriptor.
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 4]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   ```
+   var desriptorReadRequestCallback = function(clientAddress, offset) {
+      console.log(clientAddress + " requested to read desriptor's value with offset: " + offset);
+      return new tizen.GATTRequestReply(0, "0x1234");
+   };
+   server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(characteristicReadRequestCallback);
+   ```
+
+   > **Note**
+   > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` ot this descriptor.
```
Comment (djuszczakk) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
   > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` on this descriptor.
```

**Review by djuszczakk — CHANGES_REQUESTED** [review]

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 5]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
       - write about setWriteValueRequestCallback() for descriptor
  -->
 ### Responding to read and write value requests from server clients
+
+In order to react and respond to the read and write requsts from a client connected to the server running on the device, the callback functions needs to be set on the characteristics or on the descriptors.
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
In order to react and respond to the read and write value requests from a client connected to the server running on the device, the callback functions need to be set on the characteristics or on the descriptors.
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 6]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+
+To set a callback function for read or write request on a characteristic or descriptor:
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
To set a callback function for read or write value request on a characteristic or a descriptor:
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 7]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+
+1. Prepare a GATT service containing at least one characteristic or a characteristic with at least one descriptor, that can be read or write:
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
1. Prepare a GATT service containing at least one characteristic or a characteristic with at least one descriptor, that can be read or written:
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 8]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   ```
+   var exampleDescriptor = {
+      uuid: "0155",
+      /* Descriptor's permissions */
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: `readPermission` and `writePermission` are rather self-descriptive names. Please, consider removing this comment.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 9]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+      writePermission: true
+   };
+
+   var exampleCharacteristic = {
+      uuid: "0180",
+      descriptors: [exampleDescriptor],
+      /* Characteristic's properties */
+      isReadable: true,
+      isWritable: true,
+      /* Characteristic's permissions */
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: `readPermission` and `writePermission` are rather self-descriptive names. Please, consider removing this comment.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 10]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+      uuid: "0155",
+      /* Descriptor's permissions */
+      readPermission: true,
+      writePermission: true
+   };
+
+   var exampleCharacteristic = {
+      uuid: "0180",
+      descriptors: [exampleDescriptor],
+      /* Characteristic's properties */
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: `isReadable` and `isWritable` are rather self-descriptive names. Please, consider removing this comment.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 11]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+      isReadable: true,
+      isWritable: true,
+      /* Characteristic's permissions */
+      readPermission: true,
+      writePermission: true
+   };
+
+   var gattService = {
+      serviceUuid: "0955",
+      isPrimary: false,
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: Please, change this to `true` - according to the BT spec:

> A secondary service is a service that is only intended to be included from a primary service or another secondary service or other higher layer specification.

As there's no master service for gattService, it probably can't even be registered.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 12]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+      services: [],
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
      includedServices: [],
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 13]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   ```
+
+2. Register the service:
+   ```
+   var registerServiceSuccesCB = function() {
+      console.log("Service successfully registered!");
+   };
+
+   var registerServiceErrorCB = function(e) {
+      console.log("Service not registered, error: " + e);
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: 1. Please, change `e` to something more descriptive (e.g. `error`).

2.  This call will log:

> Service not registered, error: [object Object]

which is not very helpful. Please, log error's message/name or both.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 14]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   };
+
+   var server = tizen.bluetooth.getGATTServer();
+
+   server.registerService(gattService, succesCB, errorCB);
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
   server.registerService(gattService, registerServiceSuccesCB , registerServiceErrorCB);
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 15]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   ```
+
+3. Start the GATT server:
+   ```
+   var serverStartSuccessCB = function() {
+      console.log("Server started successfully!");
+   };
+
+   var serverStartErrorCB = function(e) {
+      console.error("Server didn't start, error: " + e);
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: This call will log:

    Server didn't start, error: [object Object]

which is not very helpful. Please, log error's message/name or both.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 16]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   };
+
+   server.start(serverStartSuccessCB, serverStartErrorCB);
+   ```
+
+4. To register the callback called when a client reads the value of the characteristic from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the characteristic object.
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
4. To register the callback called when a client reads the value of the characteristic from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the characteristic object:
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 17]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   ```
+   var characteristicReadRequestCallback = function(clientAddress, offset) {
+      console.log(clientAddress + " requested to read characteristic's value with offset: " + offset);
+      return new tizen.GATTRequestReply(0, "0x1234");
+   };
+   server.services[0].characteristics[0].setReadValueRequestCallback(characteristicReadRequestCallback);
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: This function takes 5 arguments:

```
void setReadValueRequestCallback(ReadValueRequestCallback readValueRequestCallback, optional SuccessCallback? successCallback,
                                 optional ErrorCallback? errorCallback, optional SuccessCallback? sendResponseSuccessCallback,
                                 optional ErrorCallback? sendResponseErrorCallback);
```

Please, define all of them to show the guide readers how to use this function properly.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 18]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
       - write about setWriteValueRequestCallback() for descriptor
  -->
 ### Responding to read and write value requests from server clients
+
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md:784]: Please, add a note (somewhere, not necessarily here), that the callbacks described here can also be registered by putting them in `BluetoothGATTServerCharacteristicInit` or `BluetoothGATTServerDescriptorInit`.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 19]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+5. To register the callback called when a client writes a value of the characteristic of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the characteristic object.
+   ```
+   var characteristicWriteRequestCallback = function(clientAddress, value, offset, replyRequired) {
+      console.log(clientAddress + " requested to write characteristic's value: " + value +
+                  " with offset: " + offset);
+      return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;
+   };
+
+   server.services[0].characteristics[0].setWriteValueRequestCallback(characteristicWriteRequestCallback);
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: Same as before - please, define and add other callbacks.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 20]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+
+   > **Note**
+   > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` on this characteristic.
+
+5. To register the callback called when a client writes a value of the characteristic of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the characteristic object.
+   ```
+   var characteristicWriteRequestCallback = function(clientAddress, value, offset, replyRequired) {
+      console.log(clientAddress + " requested to write characteristic's value: " + value +
+                  " with offset: " + offset);
+      return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: Adding the second argument in `GATTRequestReply` only makes sense for read value callbacks. Please, remove it.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 21]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+   > **Note**
+   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this characteristic.
+
+6. To register the callback called when a client reads the value of the desriptor from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the desriptor object.
+   ```
+   var desriptorReadRequestCallback = function(clientAddress, offset) {
+      console.log(clientAddress + " requested to read desriptor's value with offset: " + offset);
+      return new tizen.GATTRequestReply(0, "0x1234");
+   };
+   server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(characteristicReadRequestCallback);
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: Same as before - please, define and add other callbacks.

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 22]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+
+7. To register the callback called when a client writes a value of the desriptor of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the desriptor object.
+   ```
+   var desriptorWriteRequestCallback = function(clientAddress, value, offset, replyRequired) {
+      console.log(clientAddress + " requested to write desriptor's value: " + value +
+                  " with offset: " + offset);
+      return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;
+   };
+
+   server.services[0].characteristics[0].descriptors[0].setWriteValueRequestCallback(desriptorWriteRequestCallback);
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: Same as before - please, define and add other callbacks.

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 21 above)
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
   server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(descriptorReadRequestCallback);
```

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 23]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+
+   server.services[0].characteristics[0].setWriteValueRequestCallback(characteristicWriteRequestCallback);
+   ```
+
+   > **Note**
+   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this characteristic.
+
+6. To register the callback called when a client reads the value of the desriptor from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the desriptor object.
+   ```
+   var desriptorReadRequestCallback = function(clientAddress, offset) {
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]:
```suggestion
   var descriptorReadRequestCallback = function(clientAddress, offset) {
```

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 18 above)
Comment (pwasowski2) [reply docs/application/web/guides/connectivity/bluetooth.md:784]: Please, grep for "desriptor" and replace all occurrences with "descriptor"

File: docs/application/web/guides/connectivity/bluetooth.md [hunk 24]
```diff
@@ -781,6 +781,118 @@ To read and write a value of the Bluetooth descriptor:
...
+
+   > **Note**
+   > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` ot this descriptor.
+
+7. To register the callback called when a client writes a value of the desriptor of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the desriptor object.
+   ```
+   var desriptorWriteRequestCallback = function(clientAddress, value, offset, replyRequired) {
+      console.log(clientAddress + " requested to write desriptor's value: " + value +
+                  " with offset: " + offset);
+      return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;
```
Comment (pwasowski2) [inline docs/application/web/guides/connectivity/bluetooth.md]: Adding the second argument in GATTRequestReply only makes sense for read value callbacks. Please, remove it.

**Review by pwasowski2 — CHANGES_REQUESTED** [review]

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 11 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: OK. I'm changing this.
BTW:
I have successfully registered an only service which had `isPrimary` set to `false`.

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 18 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md:784]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 5 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 6 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 7 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 8 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 10 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 9 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 12 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 13 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 14 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 15 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 16 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 17 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: I have added paragraph with all the callbacks. I made them common for all four functions in this guide section. Please review this change.

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 20 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 19 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 1 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 23 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 21 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 4 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 2 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 24 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 22 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

File: docs/application/web/guides/connectivity/bluetooth.md (hunk 3 above)
Comment (arkadiuszpietraszek) [reply docs/application/web/guides/connectivity/bluetooth.md]: done

**Review by djuszczakk — APPROVED** [review]

**Review by pwasowski2 — APPROVED** [review]

**Pull request merged**